### DISCUSSION TITLE
Get currently typed text

### DISCUSSION THREAD

**Discussion by rawsp33d** [wezterm/wezterm#1553] [Q&A]
Is there a way to get the currently entered command, probably between the prompt and the cursor? This is the closest I have come so far:

```lua
local lines = wezterm.split_by_newlines(pane:get_lines_as_text())
print(lines[#lines])
```

In the above, the last element in `lines` is the "current" line. I _could_ take that last line, split on what I happen to know my prompt is, and take everything after the split. Unfortunately this has some major problems:
1. Most obviously, it's hacky because it all hinges on me parsing out a prompt manually, and is also tightly coupled to my specific prompt in my specific config.
2. The line also includes the "autocomplete"/"typeahead" text of my shell. Now I have never been a fan of typeahead in shells but since I couldn't find a way to disable it in the one I just switched to when I first looked, I am at least trying it out for now before digging further or opening up an issue. But regardless, to work around the typeahead text being present, I would also need to get the cursor position with `pane:get_cursor_position` and disregard everything after the cursor. Which has the side effect that the cursor being anywhere but at the end of the typed text would be problematic. It's very fair to say this is not wezterm's concern 😅
3. `pane:get_lines_as_text` supposedly takes a param for number of lines, but passing any integer there made it return nothing in my experience. This may just be a bug.

**Comment by wez** [comment]
If you're using [shell integration](https://wezfurlong.org/wezterm/shell-integration.html), your shell can markup the output using `OSC 133` to classify it as one of `Input`, `Output` and `Prompt`.

That information is exposed via the underlying `Pane::get_semantic_zones` method that is not currently exposed to lua: (this is the interface/default definition which is why it seemingly returns nothing):
https://github.com/wez/wezterm/blob/4b457da62f0b56fd7118e00d6624bd006b6691d8/mux/src/pane.rs#L359-L362

a zone has the following definition:
https://github.com/wez/wezterm/blob/4b457da62f0b56fd7118e00d6624bd006b6691d8/term/src/lib.rs#L120-L127

If that was exposed to lua, then you could find the zone that intersects with the cursor position and then you would know which portions of which lines to inspect.

Exposing the method to lua may be as simple as copying and adapting one of these, but referencing `get_semantic_zones` instead, if you feel like playing around with the code and maybe submitting a PR:

https://github.com/wez/wezterm/blob/4b457da62f0b56fd7118e00d6624bd006b6691d8/wezterm-gui/src/scripting/pane.rs#L48-L50

**Reply by rawsp33d** [reply]
I will play around with this. I used zsh for 15 years but just switched to nushell on the same day I switched to wez. It does have direct support for outputting OSC-7 nicely (`asci --osc string_terminator`) but until I add support for OSC-133 I will play around with outputting it manually, like the shell script you linked does. When I get more comfortable in this space (I'm just learning about all this semantic zones) I will maybe attempt a PR.

More importantly to me, it's nice to know you're cool with exposing more to lua. I'm enjoying the ability to script the terminal, and the more control the lua API has the better.

**Comment by rcorre** [comment]
I was trying to do the same thing. I use the `starship` prompt, which unfortunately doesn't seem to support semantic escapes right now: https://github.com/starship/starship/issues/5463

I ended up with a hacky prompt-based approach, just sharing here in case it's better than nothing for someone. `pane:get_lines_as_text(1)` does seem to be working as expected now:

```lua
  {
      key = 'Y',
      mods = 'CTRL|SHIFT',
      action = wezterm.action_callback(function(window, pane)
        local text = pane:get_lines_as_text(1)
        text = string.gsub(text, "✦ ❯ ", "")
        window:copy_to_clipboard(text, "Clipboard")
      end),
    },
```

**Reply by owenstake** [reply]
Optimizing base on you solution for powershell prompt. Auto finding the first prompt line from bottom to up. Notice wezterm will trim the tailing space at the line.

```lua
  {
      key = 'Y',
      mods = 'CTRL|SHIFT',
      action = wezterm.action_callback(function(window, pane)
          local dimentions = pane:get_dimensions()
          local height = dimentions.viewport_rows
          for i = 1,height,1 
              do 
                  local text = pane:get_lines_as_text(i)
                  local cmd = string.match(text, '^PS%s[^>]+>(.*)$')  -- non-greedy match
                  if cmd then
                      if cmd ~= "" then
                          window:copy_to_clipboard(text)
                      end
                      break
                  end
              end
      end),
  },
```

**Comment by robsonjorgefernandez** [comment]
Hi guys!

I did something similar, copy the current text from your prompt.

```lua
wezterm.on('copy-command-line-prompt', function(window, pane)

  local command = nil
  local last_line = nil
  local lines = {}
  local text = pane:get_logical_lines_as_text()

  --- Split string by '\n' and add to a table
  for line in string.gmatch(text, "([^\n]*)\n?") do
    table.insert(lines, line)
  end

  last_line = lines[#lines]
  command = string.match(last_line, "^❯ (.*)")
    or string.match(last_line, "^.+#%s+(.*)$")
    or string.match(last_line, "^.+%$%s+(.*)$")

  if command then
    if string.len(command) > 1 then
      wezterm.log_info("copy command line prompt with " .. string.len(command) .. " characters to clipboard.")
      window:copy_to_clipboard(command)
    end
  end

  -- Flash screen
  wezterm.emit('flash-terminal', window, pane)
end)
``` 
And then add to you key mapping config:
```lua
{ key = "y",  mods = "CTRL", action = action.EmitEvent "copy-command-line-prompt" }
```

It works with the following **PS1** formats, otherwise you will need to change **string.match**:

Copy everything after the character: **#**
```shell
root@16b747274e9d:/project# ls -l
```

Copy everything after the character: **$**
```shell
[📦 user@16b747274e9d:/project]$ uptime
```

Copy everything after the character: **❯**
```shell
~/R/p/gitl/r/dotfiles
❯ date
```
